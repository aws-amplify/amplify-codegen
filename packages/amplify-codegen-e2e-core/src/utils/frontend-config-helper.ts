export enum AmplifyFrontend {
  javascript = 'javascript',
  ios = 'ios',
  android = 'android',
  flutter = 'flutter'
}

export type AmplifyFrontendConfig = {
  frontendType: AmplifyFrontend;
  srcDir: string;
  modelgenDir: string;
  graphqlCodegenDir?: string;
}

export const DEFAULT_JS_CONFIG: AmplifyFrontendConfig = {
  frontendType: AmplifyFrontend.javascript,
  srcDir: 'src',
  modelgenDir: 'src/models',
  graphqlCodegenDir : 'src/graphql'
};

export const DEFAULT_ANDROID_CONFIG: AmplifyFrontendConfig = {
  frontendType: AmplifyFrontend.android,
  srcDir: 'app/src/main/res',
  modelgenDir: 'app/src/main/java',
  graphqlCodegenDir : 'app/src/main/graphql'
};

export const DEFAULT_IOS_CONFIG: AmplifyFrontendConfig = {
  frontendType: AmplifyFrontend.ios,
  srcDir: '.',
  modelgenDir: 'amplify/generated/models',
  graphqlCodegenDir : 'graphql'
};

export const DEFAULT_FLUTTER_CONFIG: AmplifyFrontendConfig = {
  frontendType: AmplifyFrontend.flutter,
  srcDir: 'lib',
  modelgenDir: 'lib/models',
};