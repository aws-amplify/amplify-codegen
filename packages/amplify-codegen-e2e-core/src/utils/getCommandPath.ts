import { execSync } from 'child_process';
/* Get the full absolute path of a command with `which`
 *
 * Windows executors on CB fail to spawn a command with custom nexpect unless the full path is used.
 */
export function getCommandPath(command: string): string {
  return execSync(`which ${command}`)
    .toString()
    .trim();
}
