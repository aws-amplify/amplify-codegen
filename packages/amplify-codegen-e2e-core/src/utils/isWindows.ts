import os from 'os';

export function isWindows(): boolean {
  return os.platform() === 'win32'
}
