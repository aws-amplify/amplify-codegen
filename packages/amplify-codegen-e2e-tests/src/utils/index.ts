import { existsSync, writeFileSync, readdirSync, mkdirSync } from "fs";
import path from 'path';

export function isNotEmptyDir(dirPath: string) : boolean {
  return existsSync(dirPath) && readdirSync(dirPath).length > 0;
}

export function generateSourceCode(projectRoot: string, srcDir: string) : string {
  const userFileData = 'This is a pre-existing file.';
  const srcCodePath = path.join(projectRoot, srcDir, 'sample.txt');
  if (!existsSync(path.dirname(srcCodePath))) {
    mkdirSync(path.dirname(srcCodePath), {recursive: true});
  }
  writeFileSync(srcCodePath, userFileData);
  return srcCodePath;
}