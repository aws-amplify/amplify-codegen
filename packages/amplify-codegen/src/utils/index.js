const getFrontEndHandler = require('./getFrontEndHandler');
const getFrontEndFramework = require('./getFrontEndFramework');
const getAppSyncAPIDetails = require('./getAppSyncAPIDetails');
const downloadIntrospectionSchema = require('./downloadIntrospectionSchema');
const getSchemaDownloadLocation = require('./getSchemaDownloadLocation');
const getIncludePattern = require('./getIncludePattern');
const getAppSyncAPIInfo = require('./getAppSyncAPIInfo');
const getAppSyncAPIInfoFromProject = require('./getAppSyncAPIInfoFromProject');
const getProjectAwsRegion = require('./getProjectAWSRegion');
const getGraphQLDocPath = require('./getGraphQLDocPath');
const downloadIntrospectionSchemaWithProgress = require('./generateIntrospectionSchemaWithProgress');
const isAppSyncApiPendingPush = require('./isAppSyncApiPendingPush');
const updateAmplifyMeta = require('./updateAmplifyMeta');
const isCodegenConfigured = require('./isCodegenConfigured');
const getSDLSchemaLocation = require('./getSDLSchemaLocation');
const switchToSDLSchema = require('./switchToSDLSchema');
const ensureIntrospectionSchema = require('./ensureIntrospectionSchema');
const { readSchemaFromFile } = require('./readSchemaFromFile');
const defaultDirectiveDefinitions = require('./defaultDirectiveDefinitions');
module.exports = {
  getAppSyncAPIDetails,
  getFrontEndHandler,
  getFrontEndFramework,
  getSchemaDownloadLocation,
  downloadIntrospectionSchema,
  downloadIntrospectionSchemaWithProgress,
  getIncludePattern,
  getAppSyncAPIInfo,
  getAppSyncAPIInfoFromProject,
  getProjectAwsRegion,
  getGraphQLDocPath,
  isAppSyncApiPendingPush,
  updateAmplifyMeta,
  isCodegenConfigured,
  getSDLSchemaLocation,
  switchToSDLSchema,
  ensureIntrospectionSchema,
  readSchemaFromFile,
  defaultDirectiveDefinitions,
};
