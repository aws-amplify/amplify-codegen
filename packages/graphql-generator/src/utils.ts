import { Language, Platform, Target, FileExtension } from './type';

export const targetToFileExtensionMap: { [key in Target]: FileExtension } = {
  javascript: 'js',
  graphql: 'graphql',
  flow: 'js',
  typescript: 'ts',
  angular: 'graphql',
};

export const platformToLanguageMap: { [key in Platform]: Language } = {
  android: 'java',
  ios: 'swift',
  flutter: 'dart',
  javascript: 'javascript',
  introspection: 'introspection',
};
