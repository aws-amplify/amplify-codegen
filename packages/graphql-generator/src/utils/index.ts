export { GraphQLStatementsFormatter } from './GraphQLStatementsFormatter';
export { statementsTargetToFileExtensionMap } from './maps';
