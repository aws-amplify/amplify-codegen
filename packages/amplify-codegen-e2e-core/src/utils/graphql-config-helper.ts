import { AmplifyFrontendConfig, AmplifyFrontend } from './frontend-config-helper';

export function constructGraphQLConfig(
    projectName: string,
    config: AmplifyFrontendConfig,
    maxDepth: number,
    region: string,
    isConfigured: boolean = false,
    apiId: string = null) : any {

    let schemaPath: string;
    let includes: [string];
    let excludes: [string] = ['./amplify/**'];
    let extensions = {
        amplify: {
            apiId: apiId,
            codeGenTarget: '',
            docsFilePath: '',
            generatedFileName: '',
            maxDepth: maxDepth ? maxDepth : 2,
            region: region
        }
    };

    switch (config.frontendType) {
      case AmplifyFrontend.android:
        schemaPath = 'app/src/main/graphql/schema.json';
        includes = ['app/src/main/graphql/**/*.graphql'];
        extensions.amplify.codeGenTarget = isConfigured ? 'android' : '';
        extensions.amplify.docsFilePath = 'app/src/main/graphql/com/amazonaws/amplify/generated/graphql';
        break;
      case AmplifyFrontend.ios:
        schemaPath = `amplify/backend/api/${projectName}/build/schema.graphql`;
        includes = ['graphql/**/*.graphql'];
        extensions.amplify.codeGenTarget = 'swift';
        extensions.amplify.docsFilePath = 'graphql';
        extensions.amplify.generatedFileName = 'API.swift';
        break;
      default:
        schemaPath = `amplify/backend/api/${projectName}/build/schema.graphql`;
        includes = ['src/graphql/**/*.js'];
        extensions.amplify.codeGenTarget = 'javascript';
        extensions.amplify.docsFilePath = 'src/graphql';
    }

    return {
        schemaPath: schemaPath,
        includes: includes,
        excludes: excludes,
        extensions: extensions
    };
  }
