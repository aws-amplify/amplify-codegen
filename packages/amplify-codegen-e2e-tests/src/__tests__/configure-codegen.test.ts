import { 
    createNewProjectDir, 
    deleteProjectDir,
    deleteProject,
    initProjectWithProfile,
    addApiWithSchema,
    addCodegen,
    DEFAULT_JS_CONFIG,
    DEFAULT_ANDROID_CONFIG,
    DEFAULT_IOS_CONFIG,
    AmplifyFrontendConfig,
    configureCodegen
  } from "amplify-codegen-e2e-core";
  import { existsSync, readFileSync } from "fs";
  import path from 'path';
  import { isNotEmptyDir, generateSourceCode } from '../utils';
  
  const frontendConfigs: AmplifyFrontendConfig[] = [
    DEFAULT_JS_CONFIG,
    DEFAULT_ANDROID_CONFIG,
    DEFAULT_IOS_CONFIG
  ];
  const schema = 'simple_model.graphql';
  const graphqlConfigFile = '.graphqlconfig.yml';
  
  describe('codegen configure tests', () => {
    let projectRoot: string;
  
    beforeEach(async () => {
      projectRoot = await createNewProjectDir('configureCodegen');
    });
  
    afterEach(async () => {
      const metaFilePath = path.join(projectRoot, 'amplify', '#current-cloud-backend', 'amplify-meta.json');
      if (existsSync(metaFilePath)) {
        await deleteProject(projectRoot);
      }
      deleteProjectDir(projectRoot);
    });

    frontendConfigs.forEach(config => {
      it(`Updating codegen configuration works as expected in ${config.frontendType} project`, async () => {
        // init project and add API category
        await initProjectWithProfile(projectRoot, { ...config });
        await addApiWithSchema(projectRoot, schema);

        // generate pre-existing user file
        const userSourceCodePath = generateSourceCode(projectRoot, config.srcDir);
        expect(existsSync(userSourceCodePath)).toBe(true);

        // add codegen
        await expect(addCodegen(projectRoot, { ...config })).resolves.not.toThrow();
        expect(existsSync(path.join(projectRoot, graphqlConfigFile))).toBe(true);
        expect(isNotEmptyDir(path.join(projectRoot, config.graphqlCodegenDir))).toBe(true);

        const addedCodegenConfiguration = readFileSync(path.join(projectRoot, graphqlConfigFile)).toString();

        // update codegen configuration
        const settings = { isCodegenConfigured: true, maxStatementDepth: 4, ...config };
        await expect(configureCodegen(projectRoot, settings)).resolves.not.toThrow();

        // pre-existing file should still exist
        expect(existsSync(userSourceCodePath)).toBe(true);
        // previously generated files should still exist
        expect(isNotEmptyDir(path.join(projectRoot, config.graphqlCodegenDir))).toBe(true);
        // graphql configuration should be updated to remove previous configuration
        const updatedCodegenConfiguration = readFileSync(path.join(projectRoot, graphqlConfigFile)).toString();
        // the codegen configuration is updated
        expect(addedCodegenConfiguration).not.toMatch(updatedCodegenConfiguration);
        // check if the updated codegen configuration is valid
        expect(updatedCodegenConfiguration).toMatchSnapshot();
      });

      it(`Adding codegen configuration works as expected in ${config.frontendType} project`, async () => {
        // init project and add API category
        await initProjectWithProfile(projectRoot, { ...config });
        await addApiWithSchema(projectRoot, schema);

        // generate pre-existing user file
        const userSourceCodePath = generateSourceCode(projectRoot, config.srcDir);
        expect(existsSync(userSourceCodePath)).toBe(true);

        // given that the codegen config file and generated statements don't exist
        expect(existsSync(path.join(projectRoot, graphqlConfigFile))).toBe(false);
        expect(isNotEmptyDir(path.join(projectRoot, config.graphqlCodegenDir))).toBe(false);

        // add default codegen configuration
        await expect(configureCodegen(projectRoot, { ...config })).resolves.not.toThrow();

        // pre-existing file should still exist
        expect(existsSync(userSourceCodePath)).toBe(true);
        // statements should be generated in expected directory
        expect(isNotEmptyDir(path.join(projectRoot, config.graphqlCodegenDir))).toBe(true);
        // graphql configuration should be added
        expect(existsSync(path.join(projectRoot, graphqlConfigFile))).toBe(true);
        // check if the added codegen configuration is valid
        expect(readFileSync(path.join(projectRoot, graphqlConfigFile)).toString()).toMatchSnapshot();
      });
    });
  });