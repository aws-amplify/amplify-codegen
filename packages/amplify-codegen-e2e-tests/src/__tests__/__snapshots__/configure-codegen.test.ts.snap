// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen configure tests Adding codegen configuration works as expected in android project 1`] = `
"projects:
  configurecodegen1eab:
    schemaPath: app/src/main/graphql/schema.json
    includes:
      - app/src/main/graphql/**/*.graphql
    excludes:
      - ./amplify/**
    extensions:
      amplify:
        codeGenTarget: ''
        generatedFileName: ''
        docsFilePath: app/src/main/graphql/com/amazonaws/amplify/generated/graphql
        region: us-west-2
        apiId: null
        maxDepth: 2
"
`;

exports[`codegen configure tests Adding codegen configuration works as expected in ios project 1`] = `
"projects:
  configurecodegen1eab:
    schemaPath: amplify/backend/api/configurecodegen1eab/build/schema.graphql
    includes:
      - graphql/**/*.graphql
    excludes:
      - ./amplify/**
    extensions:
      amplify:
        codeGenTarget: swift
        generatedFileName: API.swift
        docsFilePath: graphql
        region: us-west-2
        apiId: null
        maxDepth: 2
"
`;

exports[`codegen configure tests Adding codegen configuration works as expected in javascript project 1`] = `
"projects:
  configurecodegen1eab:
    schemaPath: amplify/backend/api/configurecodegen1eab/build/schema.graphql
    includes:
      - src/graphql/**/*.js
    excludes:
      - ./amplify/**
    extensions:
      amplify:
        codeGenTarget: javascript
        generatedFileName: ''
        docsFilePath: src/graphql
        region: us-west-2
        apiId: null
        maxDepth: 2
"
`;

exports[`codegen configure tests Updating codegen configuration works as expected in android project 1`] = `
"projects:
  configurecodegen1eab:
    schemaPath: app/src/main/graphql/schema.json
    includes:
      - app/src/main/graphql/**/*.graphql
    excludes:
      - ./amplify/**
    extensions:
      amplify:
        codeGenTarget: android
        generatedFileName: ''
        docsFilePath: app/src/main/graphql/com/amazonaws/amplify/generated/graphql
        region: us-west-2
        apiId: null
        maxDepth: 4
extensions:
  amplify:
    version: 3
"
`;

exports[`codegen configure tests Updating codegen configuration works as expected in ios project 1`] = `
"projects:
  configurecodegen1eab:
    schemaPath: amplify/backend/api/configurecodegen1eab/build/schema.graphql
    includes:
      - graphql/**/*.graphql
    excludes:
      - ./amplify/**
    extensions:
      amplify:
        codeGenTarget: swift
        generatedFileName: API.swift
        docsFilePath: graphql
        region: us-west-2
        apiId: null
        maxDepth: 4
extensions:
  amplify:
    version: 3
"
`;

exports[`codegen configure tests Updating codegen configuration works as expected in javascript project 1`] = `
"projects:
  configurecodegen1eab:
    schemaPath: amplify/backend/api/configurecodegen1eab/build/schema.graphql
    includes:
      - src/graphql/**/*.js
    excludes:
      - ./amplify/**
    extensions:
      amplify:
        codeGenTarget: javascript
        generatedFileName: ''
        docsFilePath: src/graphql
        region: us-west-2
        apiId: null
        maxDepth: 4
extensions:
  amplify:
    version: 3
"
`;
